zerver/lib/actions.py

```diff
@@ -817,6 +817,67 @@ def do_send_messages(messages):
     # intermingle sending zephyr messages with other messages.
     return already_sent_ids + [message['message'].id for message in messages]
 
+def do_send_typing_notification(notification):
+    # type: (Dict[str, Any]) -> None
+    if notification['recipient'].type == Recipient.PERSONAL:
+        notification['recipients'] = [notification['sender'],
+                                      get_user_profile_by_id(notification['recipient'].type_id)]
+    elif notification['recipient'].type == Recipient.HUDDLE:
+        # We use select_related()/only() here, while the PERSONAL case above uses
+        # get_user_profile_by_id() to get UserProfile objects from cache.  Streams will
+        # typically have more recipients than PMs, so get_user_profile_by_id() would be
+        # a bit more expensive here, given that we need to hit the DB anyway and only
+        # care about the email from the user profile.
+        fields = [
+            'user_profile__id',
+            'user_profile__email',
+            'user_profile__is_active',
+            'user_profile__realm__domain'
+        ]
+        query = Subscription.objects.select_related("user_profile", "user_profile__realm").only(*fields).filter(
+                recipient=notification['recipient'], active=True)
+        notification['recipients'] = [s.user_profile for s in query]
+    else:
+        raise ValueError('Bad recipient type')
+
+    # Only deliver the message to active user recipients
+    notification['active_recipients'] = [user_profile for user_profile in notification['recipients']
+                                         if user_profile.is_active]
+
+    event = dict(
+            type            = 'typing',
+            op              = notification['op'],
+            sender          = notification['sender'],
+            recipients      = notification['recipients'])
```

Comment:
So, Zulip's event objects need to be marshalled as JSON and then are accessed by JavaScript code, which means they shouldn't have things like Python UserProfile objects in them; that's why in `do_send_messages` we marshall all the users into dictionaries containing the user IDs and email addresses.